In [1]:
import pandas as pd

df = pd.read_csv('https://raw.githubusercontent.com/vkoul/data/main/misc/ab_bidding.csv')

print(df.shape)
print(df.head())
print(df.info())

(80, 5)
   Impression  Click  Purchase  Earning    Group
0       82529   6090       665     2311  control
1       98050   3383       315     1743  control
2       82696   4168       458     1798  control
3      109914   4911       487     1696  control
4      108458   5988       441     1544  control
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80 entries, 0 to 79
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Impression  80 non-null     int64 
 1   Click       80 non-null     int64 
 2   Purchase    80 non-null     int64 
 3   Earning     80 non-null     int64 
 4   Group       80 non-null     object
dtypes: int64(4), object(1)
memory usage: 3.3+ KB
None


In [2]:
df.groupby('Group')['Group'].value_counts()

,count
Group,
control,40
test,40


In [3]:
df.groupby('Group')[['Purchase', 'Earning']].mean()

,Purchase,Earning
Group,,
control,550.90,1908.575
test,582.05,2514.925


In [4]:
from scipy import stats

control = df[df['Group'] == 'control']['Purchase']
test = df[df['Group'] == 'test']['Purchase']

stat, p_value = stats.ttest_ind(control, test)
print(f'p-value: {p_value:.4f}')

p-value: 0.3503


In [5]:

control_earning = df[df['Group'] == 'control']['Earning']
test_earning = df[df['Group'] == 'test']['Earning']

stat_e, p_value_e = stats.ttest_ind(control_earning, test_earning)
print(f'Earning — p-value: {p_value_e:.4f}')
print(f'Control mean Earning: {control_earning.mean():.2f}')
print(f'Test mean Earning: {test_earning.mean():.2f}')

lift = (test_earning.mean() - control_earning.mean()) / control_earning.mean() * 100
print(f'Lift: {lift:.1f}%')

Earning — p-value: 0.0000
Control mean Earning: 1908.58
Test mean Earning: 2514.93
Lift: 31.8%


The difference in purchase count between average bidding and maximum bidding is not statistically significant (p=0.35) — so on its own, that metric doesn't tell us much.
But purchase count isn't the metric that matters here — revenue is. Checking Earning, maximum bidding (test) generates 31.8% more revenue than average bidding (control), and this difference is highly significant (p≈0.0000).
Recommendation: Switch to maximum bidding. It doesn't produce more purchases, but each purchase is worth significantly more — the strategy is winning higher-value opportunities, not just more of them. Revenue, not purchase count, is the right metric to make this call on.